lib/common/common/src/iterator_ext/mod.rs

```diff
@@ -42,6 +46,68 @@ pub trait IteratorExt: Iterator {
     {
         OnFinalCount::new(self, f)
     }
+
+    /// Measures the hardware usage of an iterator.
+    ///
+    /// # Arguments
+    /// - `hw_acc`: accumulator holding a counter cell
+    /// - `multiplier`: multiplies the number of iterations by this factor.
+    /// - `f`: Closure to get the specific counter to increase from the cell inside the accumulator.
+    fn measure_hw_with_acc<R>(
+        self,
+        hw_acc: HwMeasurementAcc,
+        multiplier: usize,
+        mut f: R,
+    ) -> OnFinalCount<Self, impl FnMut(usize)>
+    where
+        Self: Sized,
+        R: FnMut(&HardwareCounterCell) -> &CounterCell,
+    {
+        OnFinalCount::new(self, move |i| {
+            let hw_counter = hw_acc.get_counter_cell();
+            // Subtract 1 to not account for the latest `None` call.
```

Comment:
Does it mean that `i` in this context is not a number of values in the iterator?

If so, shouldn't it be fixed inside `OnFinalCount`? Or if it is intended, what's a true meaning of `i`? I would prefer a more explicit naming if possible.

Reply:
`i` represents how many times we called `.next()` on an iterator, which might be less than the actual amount of items in the iterator e.g. if we stopped early.
The call of `next()` that returns `None` is counted too in `i`. This behavior is documented on the `on_final_count` function of this trait. Because we don't want this measurement, we subtract `1` here.